soc: espressif: psram as shared multi heap

LucasTambor opened the pull request:
Currently, if the user wants to allocate heap on external RAM
he needs to enable CONFIG_ESP_SPIRAM and set a threshold defined
with CONFIG_ESP_HEAP_MIN_EXTRAM_THRESHOLD.

This approach requires that we re-implement k_malloc and allocate
the memory on the proper region based on the block size.

By using the shared multi heap feature the proccess of allocating
memory from external memory becomes more fluent and simple.

The attribute SMH_REG_ATTR_EXTERNAL was added to reference the
external memory.

uLipe commented on soc/espressif/esp32s3/soc.c:
```diff
@@ -211,6 +185,13 @@ void IRAM_ATTR __esp_platform_start(void)
+#if CONFIG_ESP_SPIRAM
+	/* Init Shared Multi Heap for PSRAM */
+	if (esp_psram_smh_init()) {
+		abort();
```
Question, is abort() the best approach here? PSRAM that fails in its initialization does not mean the entire system has been crashed, so a LOG_ERR instead would offer a more descriptive diagnostics, and avoid a (initially at least) cycle of rebooting when something in the PSRAM fails.

uLipe commented on soc/espressif/esp32s2/soc.c:
```diff
@@ -143,6 +119,14 @@ void __attribute__((section(".iram1"))) __esp_platform_start(void)
+
+#if CONFIG_ESP_SPIRAM
+	/* Init Shared Multi Heap for PSRAM */
+	if (esp_psram_smh_init()) {
```
I noticed the PSRAM is started at very early kernel stage without even logging or printk, it would not be better to move it to POST_KERNEL with the highest priority instead? It would better debugging experience since tracing and kernel would be available and I suppose the external heap would be consumed for POST_KERNEL components as well like wifi, camera, display or I2S.

marekmatej replied:
Provided we are sure the `.ext_ram.bss` section won't be used in PRE_KERNEL, we could do it. However, if its only merit is to inform the user, there are log messages informing user about PSRAM initialization, so I'm not sure it would be much useful.

uLipe reviewed (commented):
LGTM so far, although I have some questions.

LucasTambor committed:
soc: espressif: psram as shared multi heap

Currently, if the user wants to allocate heap on external RAM
he needs to enable CONFIG_ESP_SPIRAM and set a threshold defined
with CONFIG_ESP_HEAP_MIN_EXTRAM_THRESHOLD.

This approach requires that we re-implement `k_malloc` and allocate
the memory on the proper region based on the block size.

By using the shared multi heap feature the proccess of allocating
memory from external memory becomes more fluent and simple.

The attribute SMH_REG_ATTR_EXTERNAL was added to reference the
external memory.

Signed-off-by: Lucas Tamborrino <lucas.tamborrino@espressif.com>

LucasTambor committed:
samples: boards: esp32: spiram test

Simplifies psram test by using smh API.

Signed-off-by: Lucas Tamborrino <lucas.tamborrino@espressif.com>

LucasTambor committed:
tests: boards: esp32: Use smh in cache coex test

Simplifies test by using smh API.

Signed-off-by: Lucas Tamborrino <lucas.tamborrino@espressif.com>

epc-ake commented:
I tested the sample successfully on the `xiao_esp32s3`. 
Just needed to set `CONFIG_SPIRAM_MODE_OCT=y` manually.

sylvioalves commented on tests/boards/espressif_esp32/cache_coex/src/cache_coex.c:
```diff
@@ -11,6 +11,7 @@
+#include <zephyr/multi_heap/shared_multi_heap.h>
 
 /* definitions used in Flash & RAM operations */
 #define SPIRAM_ALLOC_SIZE               (24 * 1024)
```
Should this be retrieved from dts instead?

LucasTambor replied:
This is just the buffer size for the test. It is there since the test creation. I don't think it should be a dts related prop.

andyross reviewed (approved):
Looks clean.  This is exactly the kind of use case that multiheap was aimed at.
